common/main/java/com/couchbase/lite/AbstractDatabase.java

```diff
@@ -1576,19 +1576,34 @@ private void shutdown(boolean failIfClosed, Fn.ConsumerThrows<C4Database, LiteCo
         try {
             for (int i = 0; ; i++) {
                 verifyActiveProcesses();
+
                 if ((i >= DB_CLOSE_MAX_RETRIES) && (closeLatch.getCount() > 0)) {
                     throw new CouchbaseLiteException("Shutdown failed", CBLError.Domain.CBLITE, CBLError.Code.BUSY);
                 }
-                if (closeLatch.await(DB_CLOSE_WAIT_SECS, TimeUnit.SECONDS)) { break; }
+
+                if (closeLatch.await(DB_CLOSE_WAIT_SECS, TimeUnit.SECONDS)) {
+                    try {
+                        synchronized (getDbLock()) { onShut.accept(c4Db); }
+                        break;
+                    }
+                    catch (LiteCoreException e) {
+                        if ((e.getDomain() != C4Constants.ErrorDomain.LITE_CORE)
+                            || (e.getCode() != C4Constants.LiteCoreError.BUSY)) {
+                            throw CouchbaseLiteException.convertException(e);
+                        }
+                    }
+                }
+
+                // If we get here then, despite the fact that it appears to us that all
+                // active processes have been stopped, LiteCore has other ideas.
+                // We have no way of finding out what LiteCore thinks, other than waiting
+                // a bit and trying the again.  Since verifyActiveProcess will count down
+                // the latch, we need a new one with a count of at least 2 in order to force
+                // a wait
+                closeLatch = new CountDownLatch(2);
```

Comment:
It took me awhile to understand why setting to 2. 

As closeLatch() is an instance variable, could it be a problem for an edge case when database.close() is somehow called twice from a different thread while the first call is still waiting.

Reply:
No.  It can't be called from a different thread.  This point in the shutdown method can be reached only once, for any Database instance.  See this code at the beginning of the method:

`        synchronized (getDbLock()) {
            final boolean open = isOpenLocked();
            Log.d(DOMAIN, "Shutdown (%b, %b)", failIfClosed, open);
            if (!(failIfClosed || open)) { return; }

            c4Db = getOpenC4DbLocked();
            setC4DatabaseLocked(null);
            // mustBeOpen will now fail, which should prevent any new processes from being registered.
            
            ...
`
Only one thread at a time can be in the synchronized section.  That thread will set c4db to null.  Any subsequent thread will exit immediately
The rest of the method will be executed only once per Db

Reply:
Got it.